includes/Modules/Search_Console.php

```diff
@@ -186,23 +186,53 @@ protected function create_data_request( Data_Request $data ) {
 			case 'GET:matched-sites':
 				return $this->get_webmasters_service()->sites->listSites();
 			case 'GET:searchanalytics':
-				list ( $start_date, $end_date ) = $this->parse_date_range(
-					$data['dateRange'] ?: 'last-28-days',
-					$data['compareDateRanges'] ? 2 : 1,
-					3
-				);
+				$start_date = $data['startDate'];
+				$end_date   = $data['endDate'];
+				if ( ! strtotime( $start_date ) || ! strtotime( $end_date ) ) {
+					list ( $start_date, $end_date ) = $this->parse_date_range(
+						$data['dateRange'] ?: 'last-28-days',
+						$data['compareDateRanges'] ? 2 : 1,
+						3
+					);
+				}
 
 				$data_request = array(
-					'page'       => $data['url'],
 					'start_date' => $start_date,
 					'end_date'   => $end_date,
-					'dimensions' => array_filter( explode( ',', $data['dimensions'] ) ),
 				);
 
+				if ( ! empty( $data['url'] ) ) {
+					$data_request['page'] = $data['url'];
+				}
+
 				if ( isset( $data['limit'] ) ) {
 					$data_request['row_limit'] = $data['limit'];
 				}
 
+				$dimensions = $data['dimensions'];
+				$dimensions = array_map(
+					'trim',
+					array_filter(
+						function( $dimension ) {
+							if ( ! is_string( $dimension ) ) {
+								return false;
+							}
+
+							$dimension = trim( $dimension );
+							if ( empty( $dimension ) ) {
+								return false;
+							}
+
+							return $dimension;
+						},
+						is_array( $dimensions ) ? $dimensions : explode( ',', $dimensions )
+					)
+				);
```

Comment:
Let's put this into a `parse_report_dimensions` helper method with PHPUnit test coverage.

Also, we should maybe align this logic to be the same as https://github.com/google/site-kit-wp/pull/1810/files#r465934242? We could possibly put the utility method `parse_string_list` into the `Module` base class, what do you think?

Reply:
IMO we should avoid bringing new functionality straight into Module and its children classes because it's already like a Swiss Army knife. Such big classes hard to understand, test and work with. I wish we created more specialized classes that are responsible for a single thing and follow SOLID principles.

API requests are good canditates to be extracted into separate classes where each class is responsible for a single request handling. Then we would be able to extract common functionlity into parent classes and better cover it with unit tests.

Of course, it is out-of-scope for this ticket because it will require a lot of thinking and time to work out a good strategy for it. So, for now i'll do what you suggested and add that function to the parent Module class.

Reply:
Updated.